src/python/pants/backend/jvm/tasks/jvm_compile/rsc/rsc_compile.py

```diff
@@ -0,0 +1,603 @@
+from pants.backend.jvm.subsystems.shader import Shader
+from pants.backend.jvm.tasks.jvm_compile.compile_context import CompileContext
+from pants.backend.jvm.tasks.jvm_compile.execution_graph import Job
+from pants.backend.jvm.tasks.jvm_compile.zinc.zinc_compile import BaseZincCompile, ZincCompile
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnitLabel
+from pants.engine.fs import DirectoryToMaterialize, PathGlobs, PathGlobsAndRoot
+from pants.engine.isolated_process import ExecuteProcessRequest, FallibleExecuteProcessResult
+from pants.java.jar.jar_dependency import JarDependency
+from pants.util.contextutil import Timer
+from pants.util.dirutil import fast_relpath, safe_delete, safe_mkdir, safe_rmtree
+
+
+#
+# This is a subclass of zinc compile that uses both Rsc and Zinc to do
+# compilation.
+# It uses Rsc and the associated tools to outline scala targets. It then
+# passes those outlines to zinc to produce the final compile artifacts.
+#
+#
+logger = logging.getLogger(__name__)
+
+
+def relpathify(collection):
+  # Converts a list of paths into relpath'd paths.
+  buildroot = get_buildroot()
+  def relpath(p):
+    if p[0] != '/':
+      raise ValueError('Unexpected non-absolute path {}'.format(p))
+    return fast_relpath(p, buildroot)
+  return [relpath(c) for c in collection]
+
+
+def relpathalone(path):
+  # Converts a single path to a relpath
+  return relpathify([path])[0]
+
+
+def stdout_contents(wu):
+  if isinstance(wu, FallibleExecuteProcessResult):
+    return wu.stdout.rstrip()
+  with open(wu.output_paths()['stdout']) as f:
+    return f.read().rstrip()
+
+
+class RscCompileContext(CompileContext):
+  def __init__(self,
+               target,
+               analysis_file,
+               classes_dir,
+               rsc_mjar_file,
+               jar_file,
+               log_dir,
+               zinc_args_file,
+               sources,
+               rsc_outline_dir,
+               outline_dir):
+    super(RscCompileContext, self).__init__(target, analysis_file, classes_dir, jar_file,
+                                               log_dir, zinc_args_file, sources)
+    self.rsc_mjar_file = rsc_mjar_file
+    self.rsc_outline_dir = rsc_outline_dir
+    self.outline_dir = outline_dir
+
+  def ensure_output_dirs_exist(self):
+    safe_mkdir(os.path.dirname(self.rsc_mjar_file))
+    safe_mkdir(self.rsc_outline_dir)
+    safe_mkdir(self.outline_dir)
+
+
+class RscCompile(ZincCompile):
+  """Compile Scala and Java code to classfiles using Rsc."""
+
+  _name = 'rsc_outline' # noqa
+
+  @classmethod
+  def implementation_version(cls):
+    return super(RscCompile, cls).implementation_version() + [('RscCompile', 7)]
+
+  @classmethod
+  def register_options(cls, register):
+    super(RscCompile, cls).register_options(register)
+
+    rsc_outline_version = '0.0.0-294-d7114447'
+    scalameta_version = '4.0.0-M10'
```

Comment:
How about calling these guys `rsc_toolchain_version` and `scalameta_toolchain_version`?

Reply:
works for me.